README
David Ashley
dash@xdr.com
990207 11:24 PST

This is work in progress and might be of absolutely no use whatsoever.

This is a Bomberman clone. I bought a game called Atomic Bomberman and found
it awful, there were intolerable delays after each game in network play while
the game was downloading data (what data?) over the network. Really crappy job
they did but I liked the artwork and sound effects.

Atomic Bomberman was a DOS/win95 game. I looked under linux games and found
xblast but it was unusable for various reasons. For example its network
support just relies on xwindows networking--before entering the game you've
got to type in the host:display of each person who will be playing. Only
the main machine gets any sound. Also in the game you kept moving in whatever
direction you had last hit until you hit a wall or hit another direction.
There was no point in using the source as a starting point, there were so
many critical design flaws in that version I decided to write my own game from
scratch.

Currently the game will most likely only be playable on a LAN.

Some machine must run the matcher program. Players register their game with
the matcher machine, and joiners query the matcher machine to get a list of
active games. On xdr.com there is a matcher program running and the bomber.c
mname[] definition points to xdr.com. I tried it and it didn't work on my
own LAN through a firewall. You can try the game as is and it might work using
xdr.com as a matcher. Otherwise you'll have to run your own matcher daemon on
some machine. Assuming you have 3 machines tom, dick and harry and you want
to play a 3 player game, you would modify bomber.c's mname definiton to
be (line 47 of bomber.c)
char mname[]="tom";
Then on tom run the matcher program as a background task.
Compile and distribute duplicate copies of bomber and data/* files, then
run bomber on each machine. The menus should be self explanatory--one person
selects HOST NETWORK GAME and the other people select JOIN NETWORK GAME.

----------------- How to play  ---------------------------
You've got to blow up other players to win. Spacebar drops a bomb. Get away
and hope your enemy gets hit by the flame. The 'b' key is a 2nd control
for when you are lucky enough to pick up the bomb control--looks like a
bomb with a timer on it. When you have that the bomb won't go off until
detonated by another bomb, you are killed, or you press 'b'.

Blowing up bricks might result in prizes, most of which are good.
Skates = speed up
Bomb = allow you to have one more active bomb
flame = Increase bomb strength
turtle = makes you move very slowly
bomb with timer = controlled bomb detonation with 'b' key.
gold flame = Set flame strength to max

There isn't much point in playing the game alone (single player). In that
case the only thing to avoid is accidentally killing yourself. Big deal...
It's really a multiplayer game.

----------------- Left to do: -------------------
Come up with a better scheme for internet play, as it is any latency will
kill the game playability. Works fine on LAN though.

Figure out why matching doesn't work through a firewall...
[Specific case is 2 machines on my LAN cannot be matched by a machine that
is outside the firewall, and the firewall machine is doing IP masquerading.
Net result is matcher gets remapped IP addresses and these mean nothing
inside the LAN, only make sense when used from the outside world]

Score.

Remap movement keys not implemented.

Allow game hoster to set network game options and have them visible to players
who have joined the game.

Alternate gfx sets.

Better handling of 8 bit mode with limited palette.

Computer controlled things that can kill you.

Other bonus types that can be harmful, such as skull.

Internet play where everything is asynchronous, thus playable with a fixed
latency. As it is now slaves send to master, then master sends back to
slaves, and this data xfer must complete before the game can advance a step.
For internet this would be intolerable. I've never tried though...

Better docs, online explanations, attract mode...

Single player with computer controlled figures that must be killed/avoided.

----------------
Direct comments, complaints and questions to dash@xdr.com
This code is GPL.


ChangeLog
Fri Jun 27 16:31:51 EDT 2008
David Ashley
dashxdr@gmail.com

Turned on -Wall in Makefile, got rid of all the warnings. Made Makefile.osx
for building on OSX. Added some syntax checking for passing the "-m <host>"
command line parameter. Got rid of all the damage lists (rectangles on the
window that need to be updated) and instead just do a full screen update
whenever necessary.
---------------------------------------------------------------------------
Sun Jul 18 09:56:12 PDT 1999
Changed secondary control from ALT keys to 'b' key. On win32 the alt has
special meaning and can cause events to be dropped, so a key up event is
lost and the player gets stuck moving in some direction.

Fri Jul 16 08:23:22 PDT 1999
Finished sdl port, the sockets problem stemmed from windoze not implementing
Berkeley Sockets in a compatible manner. The program must call
WSAStartup and WSACleanup on begin and end. Also instead of close on a
socket you've got to call closesocket. This requires ugly
#if defined(__WIN32__)
...
#endif
sections in the code, which defeats the purpose of a nice platform
indpendent source tree. Blame it on M$.

Thu Jul 15 08:24:18 PDT 1999
Began converting X code to SDL code for portability. Having trouble with
sockets library under win32. Converted all the graphics rendering routines to
SDL, also dealt with the events translation. Under win32 the alt key can
cause stuck keys. Still no pause() equivalent, ie wait until the timer
interrupt occurs, but stay off the bus without eating cpu cycles.

No sound conversion yet.

Mon Jul  5 13:12:43 PDT 1999
Finally got around to dealing with old email... As per Christoph Frick's
<ridcully@futurebyte.net> suggestion I added his idea of a
"-m <matcher>" command line argument to
specify at run time the matcher to use.
Thanks Christoph!

I've bumped the version to 101.


sound.c
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <SDL_audio.h>
#include <SDL_error.h>

char dirlist[]="data";

int readsound(int num);

#define NUMSOUNDS	(sizeof(soundnames)/sizeof(char*))
#define MIXMAX 16

#define SOUND_QUIET -1

char *soundnames[] =
{
"bomb1.raw",
"power1.raw",
"death.raw",
"drop.raw",
"bomb2.raw",
"power2.raw",
};
typedef struct sample
{
	char *data;
	int len;
} sample;

#define SNDFRAGMENT 1024

sample samples[NUMSOUNDS];

int soundworking=0;
int fragment;
int soundwrite,soundread;
int *soundbuffer;
int soundbufferlen;
unsigned char sndclip[8192];

#define MAXSOUNDCOMMANDS 32
char soundcommands[MAXSOUNDCOMMANDS];
int soundtake,soundput;

int sndplaying[MIXMAX],sndposition[MIXMAX];
void fillaudio(void *udata,Uint8 *buffer,int len)
{
char com,*p;
int i,j,*ip;
int which;

	while(soundtake!=soundput)
	{
		com=soundcommands[soundtake];
		soundtake=(soundtake+1)&(MAXSOUNDCOMMANDS-1);
		if(com==SOUND_QUIET) {memset(sndposition,0,sizeof(sndposition));continue;}
		if(com<NUMSOUNDS)
		{
			for(i=0;i<MIXMAX;++i)
				if(!sndposition[i])
				{
					sndposition[i]=1;
					sndplaying[i]=com;
					break;
				}
		}
	}
	memset(soundbuffer,0,soundbufferlen);
	for(i=0;i<MIXMAX;++i)
	{
		if(!sndposition[i]) continue;
		which=sndplaying[i];
		if(sndposition[i]==samples[which].len)
		{
			sndposition[i]=0;
			continue;
		}
		p=samples[which].data;
		if(!p) continue;
		p+=len*(sndposition[i]++ -1);
		ip=soundbuffer;
		j=len;
		while(j--) *ip++ += *p++;
	}
	j=len;
	ip=soundbuffer;;
	while(j--) *buffer++ = sndclip[4096+*ip++];

}


int soundopen(void)
{
SDL_AudioSpec wanted;
int i,j;

	soundtake=soundput=0;
	memset(sndposition,0,sizeof(sndposition));
	memset(sndplaying,0,sizeof(sndplaying));
	fragment=SNDFRAGMENT<<1;
	soundbufferlen=fragment*sizeof(int);
	soundbuffer=malloc(soundbufferlen);
	if(!soundbuffer) return -2;

	memset(&wanted,0,sizeof(wanted));
	wanted.freq=22050;
	wanted.channels=2;
	wanted.format=AUDIO_U8;
	wanted.samples=fragment>>1;
	wanted.callback=fillaudio;
	wanted.userdata=0;

	if(SDL_OpenAudio(&wanted,0)<0)
	{
		fprintf(stderr,"Couldn't open audio: %s\n",SDL_GetError());
		return -1;
	}
	soundworking=1;

	for(i=0;i<8192;i++)
	{
		j=i-4096;
		sndclip[i]=j > 127 ? 255 : (j<-128 ? 0 : j+128);
	}

	for(i=0;i<NUMSOUNDS;++i)
		readsound(i);

	SDL_PauseAudio(0);
	return 0;
}
void soundclose(void)
{
	if(soundworking)
	{
		SDL_CloseAudio();
		soundworking=0;
	}
}


int readsound(int num)
{
char name[256],*p1,*p2,ch;
int i,file,size,len;
	p1=dirlist;
	for(;;)
	{
		p2=name;
		while(*p1 && (ch=*p1++)!=',')
			*p2++=ch;
		if(p2>name && p2[-1]!='/') *p2++='/';
		strcpy(p2,soundnames[num]);
		file=open(name,O_RDONLY);
		if(file>=0) break;
		if(!*p1)
		{
			samples[num].len=-1;
			return 0;
		}
	}
	size=lseek(file,0,SEEK_END);
	lseek(file,0,SEEK_SET);
	len=samples[num].len=(size+fragment-1)/fragment;
	len*=fragment;
	p1=samples[num].data=malloc(len);
	if(p1)
	{
		i=read(file,p1,size);
		if(len-size) memset(p1+size,0,len-size);
		while(size--) *p1++ ^= 0x80;
	} else
		samples[num].data=0;
	close(file);
	return 0;
}

void playsound(int n)
{
	soundcommands[soundput]=n;
	soundput=(soundput+1)&(MAXSOUNDCOMMANDS-1);
}


matcher.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/uio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/timeb.h>
#include <sys/time.h>
#include <time.h>
#include <signal.h>
#include <fcntl.h>


#define MAXMSG 256
#define PORT 5521

#define TIMETOLIVE 600 // seconds

typedef unsigned char uchar;
char logging=0;

unsigned char mesg[MAXMSG];
int lastsize;
int network;
char masterhostname[256];

#define PKT_REGISTER 16
#define PKT_ACK 24
#define PKT_INFO 40
#define PKT_QUERY 32

#define MAXMATCHES 16


struct registration
{
	uchar id;
	uchar unique[4];
	uchar password[4];
	uchar version[4];
	uchar name[16];
	uchar status;
};
struct query
{
	uchar id;
	uchar password[4];
	uchar version[4];
};

struct gamehost
{
	struct gamehost *next,*prev;
	uchar machine[4];
	uchar port[2];
	struct registration reg;
	long timeout;
};


struct gamehost *freehosts=0,*activehosts=0;



int udpsocket,myport;
struct sockaddr_in myname={0},mastername={0};
socklen_t senderlength;
struct sockaddr_in sender={0};

long longtime(void)
{
struct timeval now;
	gettimeofday(&now,0);
	return now.tv_sec;
}

char *timestr()
{
static char timestring[80];

time_t t;
int l;
	time(&t);
	strcpy(timestring,ctime(&t));
	l=strlen(timestring);
	if(l && timestring[l-1]=='\n') timestring[l-1]=0;
	return timestring;
}


int putmsg(struct sockaddr_in *toname,unsigned char *msg,int len)
{
int status;

	status=sendto(udpsocket,msg,len,0,
		(struct sockaddr *)toname,sizeof(struct sockaddr_in));
	return status;
}
void ack()
{
uchar copy[256];

	*copy=PKT_ACK;
	memmove(copy+1,mesg,lastsize);
	putmsg(&sender,copy,lastsize+1);
}

int getmsg(int seconds)
{
int size;

		lastsize=-1;
		memset(&sender,0,sizeof(sender));
		senderlength=sizeof(sender);
		if(seconds)
		{
			struct timeval timeout;
			fd_set readfds;
			int res;

			timeout.tv_sec=seconds;
			timeout.tv_usec=0;
			FD_ZERO(&readfds);
			FD_SET(udpsocket,&readfds);
			res=select(udpsocket+1,&readfds,0,0,&timeout);
			if(res<=0) return -1;
		}
		lastsize=size=recvfrom(udpsocket,mesg,MAXMSG,0,
			(struct sockaddr *)&sender,&senderlength);
		return size;
}

long longind(unsigned char *p)
{
	return (p[0]<<24L) | (p[1]<<16L) | (p[2]<<8) | p[3];
}
short shortind(unsigned char *p)
{
	return (p[0]<<8L) | p[1];
}


void openport(int portwant)
{
int status;

	myport=portwant;
	udpsocket=socket(PF_INET,SOCK_DGRAM,IPPROTO_UDP);
	if(udpsocket==-1)
	{
		perror("socket()");
		exit(1);
	}
	memset(&myname,0,sizeof(myname));
	myname.sin_family=AF_INET;
	myname.sin_addr.s_addr=htonl(INADDR_ANY);
	myname.sin_port=htons(myport);

	status=bind(udpsocket,(struct sockaddr *) &myname,sizeof(myname));
	if(status==-1)
	{
		perror("bind()");
		exit(1);
	}
}

#define PERBLOCK 512

struct gamehost *newhost()
{
struct gamehost *h,*block;
int i;

	if(!freehosts)
	{
		block=malloc(sizeof(struct gamehost)*PERBLOCK);
		if(!block) return 0;
		freehosts=block;
		i=PERBLOCK-1;
		while(i--)
		{
			block->next=block+1;
			++block;
		}
		block->next=0;
	}
	h=freehosts;
	freehosts=freehosts->next;
	memset(h,0,sizeof(struct gamehost));
	return h;
}

void freehost(struct gamehost *h)
{
	h->next=freehosts;
	freehosts=h;
}

struct gamehost *findmatch(struct registration *key)
{
struct gamehost *h;
	h=activehosts;
	while(h)
	{
		if(!memcmp(&h->reg,key,sizeof(struct registration)-1))
			return h;
		h=h->next;
	}
	return 0;
}



void insert(struct gamehost *h)
{
	if(activehosts)
	{
		h->next=activehosts;
		h->prev=activehosts->prev;
		activehosts->prev=h;
		activehosts=h;
	} else
	{
		h->next=h->prev=0;
		activehosts=h;
	}
}
void delete(struct gamehost *h)
{
	if(h->prev)
		h->prev->next=h->next;
	else
		activehosts=h->next;
	if(h->next)
		h->next->prev=h->prev;
	freehost(h);
}
void doreg()
{
struct registration *new;
struct gamehost *match;
long now;

	new=(struct registration *)mesg;
	match=findmatch(new);
	if(logging)
	{
		unsigned addr=ntohl(sender.sin_addr.s_addr);
		unsigned short port=ntohs(sender.sin_port);
		printf("reg  :%s:%d.%d.%d.%d:%d  %c%lx '%s'\n",timestr(),
			(addr>>24)&255,(addr>>16)&255,(addr>>8)&255,addr&255,port,
			new->status ? '+' : '-',(long)match,new->name);
		fflush(stdout);
	}
	if(!match && !new->status) {ack();return;}
	if(match && new->status) {ack();return;}
	if(!match && new->status)
	{
		match=newhost();
		if(!match) return; // No memory, what can we do?
		memmove(match->machine,&sender.sin_addr.s_addr,4);
		memmove(match->port,&sender.sin_port,2);
		match->reg=*new;
		now=longtime();
		match->timeout=now+TIMETOLIVE;
		ack();
		insert(match);
		return;
	} else // match && !new->status
	{
		delete(match);
		ack();
		return;
	}
}

void doquery()
{
uchar *password;
uchar *version;
struct gamehost *h;
uchar response[2048],*rput,*countersave;
int counter;

	if(logging)
	{
		unsigned addr=ntohl(sender.sin_addr.s_addr);
		unsigned short port=ntohs(sender.sin_port);
		printf("query:%s:%d.%d.%d.%d:%d\n",timestr(),
			(addr>>24)&255,(addr>>16)&255,(addr>>8)&255,addr&255,port);
		fflush(stdout);
	}
	password=mesg+1;
	version=mesg+5;
	h=activehosts;
	rput=response;
	*rput++=PKT_INFO;
	memmove(rput,password,4);
	rput+=4;
	memmove(rput,version,4);
	rput+=4;
	countersave=rput;
	*rput++=0;
	*rput++=0;
	counter=0;

	while(h)
	{
		if(!memcmp(password,h->reg.password,4) &&
			!memcmp(version,h->reg.version,4) && counter<MAXMATCHES)
		{
			++counter;
			memmove(rput,h->reg.unique,4);
			rput+=4;
			memmove(rput,h->machine,4);
			rput+=4;
			memmove(rput,h->port,2);
			rput+=2;
			memmove(rput,h->reg.name,sizeof(h->reg.name));
			rput+=sizeof(h->reg.name);
		}
		h=h->next;
	}
	*countersave++=counter>>8;
	*countersave++=counter;
	putmsg(&sender,response,rput-response);

}

void purge(long cutoff)
{
struct gamehost *h,*h2;
	h=activehosts;
	while(h)
	{
		h2=h;
		h=h->next;
		if(cutoff-h2->timeout>0)
		{
			delete(h2);
		}
	}
}


int main(int argc,char **argv)
{
int i;
int want;
int size;
long purgetime;
long now;

	want=PORT;
	if(argc>1)
	{
		for(i=1;i<argc;++i)
			if(!strncmp(argv[i],"-p",2))
			{
				if(strlen(argv[i])>2) want=atoi(argv[i]+2);
				else if(i+1<argc) want=atoi(argv[i+1]);
			}
	}
	freehosts=0;
	openport(want);
	purgetime=longtime()+TIMETOLIVE;
	for(;;)
	{
		size=getmsg(10);
		if(size>=1)
			switch(*mesg)
			{
			case PKT_REGISTER:
				if(size<sizeof(struct registration)) continue;
				doreg();
				break;
			case PKT_QUERY:
				if(size<sizeof(struct query)) continue;
				doquery();
				break;
			}
		now=longtime();
		if(now-purgetime>0) // avoid year 203x bug...
		{
			purge(purgetime);
			purgetime+=TIMETOLIVE;
		}
	}
}


bomber.h
#ifndef BOMBER_H
#define BOMBER_H

#include "SDL.h"

typedef unsigned char uchar;
extern int xcolors[256];
#define NUMGFX 3
#define IXSIZE 640
#define IYSIZE 480


#define MYF1 0x180
#define MYF2 0x181
#define MYF3 0x182
#define MYF4 0x183
#define MYF5 0x184
#define MYF6 0x185
#define MYF7 0x186
#define MYF8 0x187
#define MYF9 0x188
#define MYF10 0x189
#define MYLEFT 0x190
#define MYRIGHT 0x191
#define MYUP 0x192
#define MYDOWN 0x193
#define MYPAGEUP 0x194
#define MYPAGEDOWN 0x195
#define MYHOME 0x196
#define MYEND 0x197
#define MYALTL 0x198
#define MYALTR 0x199

#define MYDELETE 0x7f

#define MYSHIFTED 0x40
#define MYALTED 0x200


typedef struct gfxset
{
	uchar gs_colormap[768];
	uchar gs_inout[256];
	uchar *gs_pic;
	int gs_xsize;
	int gs_ysize;
} gfxset;

typedef struct figure
{
	int	xsize,ysize;
	int	xdelta,ydelta;
	uchar *graphics;
} figure;

typedef struct solid
{
	int	xsize,ysize;
	uchar	*graphics;
} solid;

typedef struct player
{
	struct player *next;
	int xpos,ypos;
	int flags;
	int abilities;
	int speed;
	int bombsused;
	int bombsavailable;
	int flamelength;
	int *at;
	int figcount;
	int doing;
	int action;
	int color;
	int controller;
	int fixx,fixy;
	figure *figure;
} player;

#define FLG_CONTROL 1
#define FLG_DEAD 2



typedef struct sprite
{
	int	flags;
	int	xpos,ypos;
	figure	*fig;
} sprite;

typedef struct damage
{
	int xpos,ypos;
	int xsize,ysize;
} damage;

typedef struct list
{
	void *next;
} list;

typedef struct bomb
{
	struct bomb *next;
	int type;
	int xpos,ypos;
	int px,py;
	int power;
	int timer;
	int figcount;
	player *owner;
} bomb;


#define BOMB_OFF 0
#define BOMB_NORMAL 1
#define BOMB_CONTROLLED 2
#define BOMBLIFE 60
#define FLAMELIFE 15
#define DECAYLIFE 15

typedef struct flame
{
	struct flame *next;
	int xpos,ypos;
	int px,py;
	int timer;
	int lurd;
	player *owner;
} flame;

#define FL_UP 2
#define FL_DOWN 8
#define FL_LEFT 1
#define FL_RIGHT 4

typedef struct brickdecay
{
	struct brickdecay *next;
	int xpos,ypos;
	int px,py;
	int timer;
} brickdecay;

typedef struct generic
{
	struct generic *next;
	int xpos,ypos;
	int px,py;
	int timer;
	void (*process)();
	void (*draw)();
	void *ptr1,*ptr2;
	int data1,data2;
} generic;

typedef struct bonustile
{
	struct bonustile *next;
	int xpos,ypos;
	int px,py;
	int type;
}bonustile;
#define TILE_NONE -1
#define TILE_BOMB 5
#define TILE_FLAME 2
#define TILE_GOLDFLAME 7
#define TILE_CONTROL 9
#define TILE_SKATES 4
#define TILE_TURTLE 14


#define ACT_INVALID 0x88
#define ACT_NONE  0
#define ACT_UP    1
#define ACT_DOWN  2
#define ACT_LEFT  3
#define ACT_RIGHT 4
#define ACT_ENTER 5
#define ACT_QUIT 6
#define ACT_MASK 0x1f
#define ACT_PRIMARY 0x40
#define ACT_SECONDARY 0x80


#define FIELD_EMPTY 0
#define FIELD_BORDER 1
#define FIELD_BOMB 2
#define FIELD_BRICK 3
#define FIELD_FLAME 4
#define FIELD_EXPLODING 5
#define FIELD_BONUS 6

#define CODE_CONT 0
#define CODE_QUIT 1
#define CODE_ALLDEAD 2



#define MAXTHINGS 500
#define MAXSETS 8
#define MAXSPRITES 128
#define MAXDAMAGES 512
#define MAXBOMBSDETONATED 32

extern char exitflag;

extern player players[];
extern sprite sprites[];
extern gfxset gfxsets[NUMGFX];
extern uchar needwhole;
extern figure walking[MAXSETS][60];
extern damage damages[];
extern bomb bombs[];

extern void centerx(player *pl);
extern void centery(player *pl);
extern void dropbomb(player *pl,int px,int py,int type);
extern void adddetonate(bomb *bmb);
extern void flameshaft(player *owner,int px,int py,int dx,int dy,int power);
extern void detonatebomb(bomb *bmb);
extern void initlist(void *first,int size,int num);
extern void freeentry(void *entry);
extern void addtail(void *header,void *entry);
extern void delink(void *header,void *entry);
extern void *allocentry();
extern void adddecay(int px,int py);
extern void trybonus(int px,int py);
extern void addbonus(int px,int py,int type);
extern void deletebonus(bonustile *bonus);
extern void queuesequence(int xpos,int ypos,figure *fig,int count);
extern void playonce(generic *gen);
extern void processgenerics(void);
extern void drawgenerics(void);
extern void drawgeneric(generic *gen);
extern void killplayer(player *pl);
extern void adddeath(player *pl);
extern void drawbigstring(int xpos,int ypos,char *str);
extern void bigscrprintf(char *str, ...);
extern void drawstring(int xpos,int ypos,char *str);
extern void scrprintf(char *str, ...);
extern int textx,texty,fontxsize,fontysize;
extern void texthome(void);

extern unsigned char field[32][32];
extern void *info[32][32];
extern unsigned char singleoptions[10];
extern unsigned char gameoptions[10];
extern int getmsg(int);

#define GO_DENSITY 0
#define GO_FLAMES 1
#define GO_BOMBS 2
#define GO_GENEROSITY 3

// network packet types
// slave -> master packets
#define PKT_MYDATA 0 // 4 bytes unique #,4 bytes frame #, 1 byte data
#define PKT_JOIN 1 // 4 bytes unique #,16 bytes name
// master -> slave packets
#define PKT_INVITE 9 // 4 bytes unique #, any # of 1:slot,16:name sets (-1 end)
#define PKT_BEGIN 10 // clone of INVITE
#define PKT_STEP 11 // 4 bytes unique #, 4 bytes frame #, 8 bytes ACT_*
#define PKT_QUIT 12 // 4 bytes unique #
// master -> matcher packets
#define PKT_REGISTER 16 // 4:unique #,4:pword hash,4:version #,16:name, 1:status
// matcher -> master packets
#define PKT_ACK 24 // perfect copy of packet received
// slave -> matcher packets
#define PKT_QUERY 32 // 4 bytes password hash
// matcher -> slave packets
#define PKT_INFO 40 // 4: pword hash, 2: count,#(4:unique,4:IP,2:port,16:name)

// all bytes stored MSB first

/*
game startup:
<master and matcher>
Master: send REGISTER to matcher with optional password, wait for ack. If
  timout, resend.
matcher: Wait for REGISTER packet, when received maintain database. respond
  to sender with ACK. REGISTER packet can close a game also. The REGISTER
  packet sent by the master has a unique word to be used to avoid confusion.
  REGISTER packet also contains a game version #

After master registers game and receives ACK, just waits for slaves to contact.

<slave and matcher>
slave: send QUERY to matcher with optional password, wait for INFO, if timeout,
  resend.
matcher: respond to QUERY with INFO packet. matcher need not maintain any
  database for slave requests. INFO packet contains IP addr and port for each
  master machine that matches the QUERY spec (ALL or password). Only a
  certain MAX # of entries are sent if there are too many to choose from.

<slave and master>
slave: send JOIN to master, wait for INVITE. If timeout, resend. JOIN packet
  contains the unique word the master created. JOIN also contains username.
master: Respond to JOIN with INVITE. INVITE contains unique word from JOIN
  packet. INVITE either contains NO meaning game no longer exists or is closed
  or player is not invited. IF yes, INVITE contains info on other players
  already in the game (username and slot # for each). Master allocates the
  slots and avoids confusion based on IP addr and port #. INVITE also contains
  game options structure. Whenever a new player JOINS and is admitted, master
  sends updated INVITE packets to everyone already in the JOIN list. Whenever
  master changes game options, master sends out another set of INVITES

Duplicate JOINS are answered with updated INVITE but nothing changes as far
as allocation.

Master player launches game after he's satisfied everyone has joined.

Master sends BEGIN packet to everyone. BEGIN is identical to INVITE except
that the data is final. Slave must respond with its first MYDATA packet with
frame # of 0. If master times out waiting, master sends duplicate BEGIN to
wayward slaves. Once master has received MYDATA from everyone, game starts.

Within game slave sends MYDATA to master and waits for STEP packet. If
timeout, slave sends duplicate MYDATA.

If master times out waiting for a slave's MYDATA, slave gets dropped. MYDATAs
received will be answered with PKT_QUIT.

*/

int soundopen(void);
void soundclose(void);
void playsound(int n);




#endif // BOMBER_H


